src/coprocessor/dag/expr/builtin_string.rs

```diff
@@ -109,6 +109,29 @@ impl ScalarFunc {
         }
         Ok(Some(Cow::Owned(output)))
     }
+    #[inline]
+    pub fn concat_ws<'a, 'b: 'a>(
+        &'b self,
+        ctx: &mut EvalContext,
+        row: &'a [Datum],
+    ) -> Result<Option<Cow<'a, [u8]>>> {
```

Comment:
Is it better to do it like this?

1. parse the separator first
2. parse the rest children one by one and collect them together 
3. use `join()` from `std::slice::SliceConcatExt` to join them together in the end

Reply:
Yes! Because  I  study the rust-lang for the first time , so  I  don't know to use the most appropriate method. Thank you very much! I will try the new code.

Reply:
How about this:

```rust
        let mut sep: Vec<u8> = Vec::new();
        let mut output: Vec<u8> = Vec::new();
        for (i, expr) in self.children.iter().enumerate() {
            let input = try_opt!(expr.eval_string(ctx, row));
            match i {
                0 => sep = input.to_vec(),
                1 => output = input.to_vec(),
                _ => {
                    output.extend_from_slice(&sep);
                    output.extend_from_slice(&input);
                }
            }
        }
        Ok(Some(Cow::Owned(output)))
```

Reply:
Yes, it is a good idea.  It is the first way i considered , but theretofore i didn't know how to get the iterator.  Thank you!